cpp/dolfinx/common/MPI.h

```diff
@@ -244,22 +244,24 @@ dolfinx::MPI::neighbor_all_to_all(MPI_Comm neighbor_comm,
   MPI_Dist_graph_neighbors_count(neighbor_comm, &indegree, &outdegree,
                                  &weighted);
 
-  // Get receive sizes
-  std::vector<int> send_sizes(outdegree, 0);
-  std::vector<int> recv_sizes(indegree);
+  // Allocate memory (add '1' to handle empty as OpenMPI fails for null pointers
+  std::vector<int> send_sizes(outdegree + 1, 0);
+  std::vector<int> recv_sizes(indegree + 1);
   std::adjacent_difference(std::next(send_data.offsets().begin()),
                            send_data.offsets().end(), send_sizes.begin());
+  // Get receive sizes
   MPI_Neighbor_alltoall(send_sizes.data(), 1, MPI::mpi_type<int>(),
                         recv_sizes.data(), 1, MPI::mpi_type<int>(),
                         neighbor_comm);
 
   // Work out recv offsets
-  std::vector<int> recv_offsets(recv_sizes.size() + 1);
+  std::vector<int> recv_offsets(indegree + 1);
```

Comment:
This seems wrong as recv_offsets is now same size as recv_sizes. Better to stick with reserve() - much clearer what is going on.

Reply:
Using `reserve` for this purpose is not good. It's making assumptions on how `std::vector` is implemented.

Reply:
Resizing to an incorrect size is not good either. The problem is in OpenMPI, and the assumptions it makes.

Reply:
Anyway, the partial_sum below now looks like it oversteps the end of recv_offsets

Reply:
Fixed.